scheds/rust/scx_rustlite/src/main.rs

```diff
@@ -0,0 +1,480 @@
+        let nr_cpus_online = libbpf_rs::num_possible_cpus().unwrap() as u64;
+
+        // Set scheduler options (defined in the BPF part).
+        skel.bss_mut().usersched_pid = pid;
+        skel.rodata_mut().slice_ns = opts.slice_us * 1000;
+        skel.rodata_mut().switch_partial = opts.partial;
+        skel.rodata_mut().debug = opts.debug;
+
+        // Attach BPF scheduler.
+        let mut skel = skel.load().context("Failed to load BPF program")?;
+        skel.attach().context("Failed to attach BPF program")?;
+        let struct_ops = Some(
+            skel.maps_mut()
+                .rustlite()
+                .attach_struct_ops()
+                .context("Failed to attach struct ops")?,
+        );
+        info!("{} scheduler attached", SCHEDULER_NAME);
+
+        // Return scheduler object.
+        Ok(Self {
+            skel,
+            task_pool,
+            task_map,
+            min_vruntime,
+            nr_cpus_online,
+            struct_ops,
+        })
+    }
+
+    // Read exit code from the BPF part.
+    fn read_bpf_exit_kind(&mut self) -> i32 {
+        unsafe { std::ptr::read_volatile(&self.skel.bss().exit_kind as *const _) }
+    }
+
+    // Called on exit to get exit code and exit message from the BPF part.
+    fn report_bpf_exit_kind(&mut self) -> Result<()> {
+        match self.read_bpf_exit_kind() {
+            0 => Ok(()),
+            etype => {
+                let cstr = unsafe { CStr::from_ptr(self.skel.bss().exit_msg.as_ptr() as *const _) };
+                let msg = cstr
+                    .to_str()
+                    .context("Failed to convert exit msg to string")
+                    .unwrap();
+                info!("BPF exit_kind={} msg={}", etype, msg);
+                Ok(())
+            }
+        }
+    }
+
+    // Update task's vruntime based on the information collected from the kernel part.
+    fn update_enqueued(
+        task_info: &mut TaskInfo,
+        sum_exec_runtime: u64,
+        weight: u64,
+        min_vruntime: u64,
+    ) {
+        // Add cputime delta normalized by weight to the vruntime (if delta > 0).
+        if sum_exec_runtime > task_info.sum_exec_runtime {
+            let delta = (sum_exec_runtime - task_info.sum_exec_runtime) / weight;
+            task_info.vruntime += delta;
+        }
+        // Make sure vruntime is moving forward (> current minimum).
+        if min_vruntime > task_info.vruntime {
+            task_info.vruntime = min_vruntime;
+        }
+        // Update total task cputime.
+        task_info.sum_exec_runtime = sum_exec_runtime;
+    }
+
+    // Drain all the tasks from the queued list, update their vruntime (Self::update_enqueued()),
+    // then push them all to the task pool (doing so will sort them by their vruntime).
+    fn drain_queued_tasks(&mut self) {
+        let maps = self.skel.maps();
+        let queued = maps.queued();
+
+        loop {
```

Comment:
Not a blocker: Ideally, we shouldn't be dumping the whole queue into the dispatcher because then the userspace scheduler loses the ability to insert tasks that wake up later but at an a lot higher priority before the waiting low pri ones. As implemented, how much ordering power the userland scheduler has is rather accidental - it depends on how often it gets to dump the queue content.

This exposes a weakness in userspace scheduling in that the more immediate the scheduling decisions are made, the higher the overhead becomes as we'd then have to wake up the userlevel thread whenever a CPU becomes idle which then may have to stall a bit until the userspace scheduler makes a decision and so on. Off the top of my head, there are a couple ways to address them:

1. Have a reasonable batch size (probably a function of number of CPUs and slice duration) so that we don't dump the whole queue which can be pretty big at once but also don't have to wakeup the userland thread on every context switch.
2. Have some ordering on the BPF dispatch side too so that userland can retroactively insert higher pri tasks after other tasks have been dispatched. e.g. Instead of using DSQ_GLOBAL use a vtime ordered custom DSQ. It does incur double rbtree queueing overhead but I doubt that would matter in this case.

Anyways, something to mull over as this gets developed.